File: app/src/main/java/dev/favourdevlabs/cleanthes/autofill/StructureParser.java
package dev.favourdevlabs.cleanthes.autofill;

import android.app.assist.AssistStructure;
import android.text.InputType;
import android.view.View;
import android.view.autofill.AutofillId;

public class StructureParser {

    // This is what we're trying to find after parsing
    public static class ParsedFields {
        public AutofillId usernameId;  // the AutofillId of the email/username field
        public AutofillId passwordId;  // the AutofillId of the password field
        public String packageName;     // which app owns this screen e.g. "com.google.android.gm"
        public String webDomain;       // if it's a browser, the actual site domain
    }

    public static ParsedFields parse(AssistStructure structure) {
        ParsedFields result = new ParsedFields();

        // An AssistStructure can have multiple windows (e.g. a dialog over an activity)
        // We iterate all of them
        int windowCount = structure.getWindowNodeCount();
        for (int i = 0; i < windowCount; i++) {
            AssistStructure.WindowNode windowNode = structure.getWindowNodeAt(i);

            // The window title is formatted as "packageName/ActivityName"
            // Split on "/" to get just the package name
            String title = windowNode.getTitle() != null
                    ? windowNode.getTitle().toString()
                    : "";
            if (title.contains("/")) {
                result.packageName = title.split("/")[0];
            }

            // Now walk the entire view tree starting from the root
            traverseNode(windowNode.getRootViewNode(), result);
        }

        return result;
    }

    private static void traverseNode(AssistStructure.ViewNode node, ParsedFields result) {
        if (node == null) return;

        // If this node is inside a browser WebView, it has a webDomain
        if (node.getWebDomain() != null) {
            result.webDomain = node.getWebDomain();
        }

        // --- Primary detection: autofillHints ---
        // This is the reliable way. App developers label their fields.
        String[] hints = node.getAutofillHints();
        if (hints != null) {
            for (String hint : hints) {
                if (isUsernameHint(hint)) {
                    result.usernameId = node.getAutofillId();
                }
                if (isPasswordHint(hint)) {
                    result.passwordId = node.getAutofillId();
                }
            }
        }

        // --- Fallback detection: inputType ---
        // Many apps don't set autofillHints. Check the keyboard type instead.
        // A password field will have TYPE_TEXT_VARIATION_PASSWORD set.
        if (result.passwordId == null) {
            int inputType = node.getInputType();
            boolean isPassword =
                    (inputType & InputType.TYPE_TEXT_VARIATION_PASSWORD) != 0
                    || (inputType & InputType.TYPE_TEXT_VARIATION_WEB_PASSWORD) != 0
                    || (inputType & InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD) != 0;

            if (isPassword) {
                result.passwordId = node.getAutofillId();
            }
        }

        // --- Fallback detection: field name/hint text ---
        // Some apps use hint text like "Enter email" — check that too
        if (result.usernameId == null) {
            CharSequence hint = node.getHint();
            if (hint != null) {
                String hintStr = hint.toString().toLowerCase();
                if (hintStr.contains("email")
                        || hintStr.contains("username")
                        || hintStr.contains("user name")
                        || hintStr.contains("phone")) {
                    result.usernameId = node.getAutofillId();
                }
            }
        }

        // Recurse into children
        for (int i = 0; i < node.getChildCount(); i++) {
            traverseNode(node.getChildAt(i), result);
        }
    }

    private static boolean isUsernameHint(String hint) {
        return hint.equalsIgnoreCase(View.AUTOFILL_HINT_USERNAME)
                || hint.equalsIgnoreCase(View.AUTOFILL_HINT_EMAIL_ADDRESS)
                || hint.equalsIgnoreCase(View.AUTOFILL_HINT_PHONE)
                || hint.equalsIgnoreCase("email")
                || hint.equalsIgnoreCase("username");
    }

    private static boolean isPasswordHint(String hint) {
        return hint.equalsIgnoreCase(View.AUTOFILL_HINT_PASSWORD)
                || hint.equalsIgnoreCase("password")
                || hint.equalsIgnoreCase("current-password");
    }
}

File: feature/autofill/src/main/java/dev/favourdevlabs/cleanthes/feature/autofill/StructureParser.kt
package dev.favourdevlabs.cleanthes.feature.autofill

import android.app.assist.AssistStructure
import android.text.InputType
import android.view.View
import android.view.autofill.AutofillId

object StructureParser {

    data class ParsedFields(
        var usernameId: AutofillId? = null,
        var passwordId: AutofillId? = null,
        var repeatPasswordId: AutofillId? = null,
        var packageName: String? = null,
        var webDomain: String? = null,
    )

    private data class Candidate(
        val id: AutofillId,
        val isFocused: Boolean,
        val tier: Int, // 1 = autofillHints, 2 = inputType, 3 = hint text
    )

    private class Accumulator {
        val usernames = mutableListOf<Candidate>()
        val passwords = mutableListOf<Candidate>()
        var packageName: String? = null
        var webDomain: String? = null
    }

    fun parse(structure: AssistStructure): ParsedFields {
        val acc = Accumulator()
        for (i in 0 until structure.windowNodeCount) {
            val windowNode = structure.getWindowNodeAt(i)
            windowNode.title?.toString()?.let { title ->
                if (title.contains("/")) acc.packageName = title.split("/")[0]
            }
            traverseNode(windowNode.rootViewNode, acc)
        }
        return resolve(acc)
    }

    private fun traverseNode(
        node: AssistStructure.ViewNode?,
        acc: Accumulator,
    ) {
        if (node == null) return

        node.webDomain?.let { acc.webDomain = it }

        if (isSupportedInput(node)) {
            classifyNode(node, acc)
        }

        for (i in 0 until node.childCount) traverseNode(node.getChildAt(i), acc)
    }

    private fun isSupportedInput(node: AssistStructure.ViewNode): Boolean =
        node.className == "android.widget.EditText" ||
            node.className == "android.widget.AutoCompleteTextView"

    private fun classifyNode(
        node: AssistStructure.ViewNode,
        acc: Accumulator,
    ) {
        val id = node.autofillId ?: return
        val focused = node.isFocused

        // --- Tier 1: explicit autofillHints ---
        var matchedTier1 = false
        node.autofillHints?.forEach { hint ->
            if (isUsernameHint(hint)) {
                acc.usernames.add(Candidate(id, focused, tier = 1))
                matchedTier1 = true
            }
            if (isPasswordHint(hint)) {
                acc.passwords.add(Candidate(id, focused, tier = 1))
                matchedTier1 = true
            }
        }
        if (matchedTier1) return

        // --- Tier 2: inputType ---
        val inputType = node.inputType
        val isPasswordType =
            (inputType and InputType.TYPE_TEXT_VARIATION_PASSWORD) != 0 ||
                (inputType and InputType.TYPE_TEXT_VARIATION_WEB_PASSWORD) != 0 ||
                (inputType and InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD) != 0
        if (isPasswordType) {
            acc.passwords.add(Candidate(id, focused, tier = 2))
            return
        }

        // --- Tier 3: hint text (heuristic, weakest signal — no "phone") ---
        node.hint?.toString()?.lowercase()?.let { hintStr ->
            if (hintStr.contains("email") ||
                hintStr.contains("username") ||
                hintStr.contains("user name")
            ) {
                acc.usernames.add(Candidate(id, focused, tier = 3))
            }
        }
    }

    private fun isUsernameHint(hint: String): Boolean =
        hint.equals(View.AUTOFILL_HINT_USERNAME, ignoreCase = true) ||
            hint.equals(View.AUTOFILL_HINT_EMAIL_ADDRESS, ignoreCase = true) ||
            hint.equals(View.AUTOFILL_HINT_PHONE, ignoreCase = true) ||
            hint.equals("email", ignoreCase = true) ||
            hint.equals("username", ignoreCase = true)

    private fun isPasswordHint(hint: String): Boolean =
        hint.equals(View.AUTOFILL_HINT_PASSWORD, ignoreCase = true) ||
            hint.equals("password", ignoreCase = true) ||
            hint.equals("current-password", ignoreCase = true)

    private val candidatePriority = compareByDescending<Candidate> { it.isFocused }.thenBy { it.tier }

    private fun resolve(acc: Accumulator): ParsedFields {
        val bestUsername = acc.usernames.sortedWith(candidatePriority).firstOrNull()
        val sortedPasswords = acc.passwords.sortedWith(candidatePriority)
        val bestPassword = sortedPasswords.firstOrNull()
        val repeatPassword = sortedPasswords.drop(1).firstOrNull { it.id != bestPassword?.id }

        return ParsedFields(
            usernameId = bestUsername?.id,
            passwordId = bestPassword?.id,
            repeatPasswordId = repeatPassword?.id,
            packageName = acc.packageName,
            webDomain = acc.webDomain,
        )
    }
}

